### PR TITLE
Implement EIP-712 Typed Data

### PR DESCRIPTION
**Pull request by peterbroadhurst**
### Summary

- Clean Apache 2.0 implementation of https://eips.ethereum.org/EIPS/eip-712
    - Initial implementation believed to contain all core features
- Re-uses the ABI encoding libraries in this repo wherever possible, for consistent encoding/parsing behavior
    - Had to expose a couple of new side-doors to the ABI package to do this - other consumers could use these too
- Operation compared with ISC License reference implementation in TypeScript:
    - See https://github.com/MetaMask/eth-sig-util/blob/main/src/sign-typed-data.ts
    - See PR https://github.com/ethereum/EIPs/pull/2190 against standard (never merged)
- Implements `V4` derivation of the behavior, as adopted by Metamask and OpenZeppelin communities
    - Most questions in the spec resolved through behavior of above reference implementation
- Extends the `ethsigner.Wallet` interface with `ethsigner.WalletTypedData`
    - Avoids existing wallet implementations being forced to change
- Extends the `fswallet.Wallet` implementation to include `SignTypedDataV4`
- Contains some general maintenance on the repo
    - Updates `firefly-common`
    - Updates to Go 1.20
    - Updates dependencies
- Provides mapping of ABI types to EIP-712 type arrays
    - Feels like the tooling around EIP-712 requires quite a bit of hand-spinning for the Solidity developer, and I was thinking this could take away at least one potential source of mistakes, and remove a 3rd artifact to maintain (Solidity struct/ABI, Solidity EIP-712 signing construct on-chain, JSON EIP-712 type array off-chain). Maybe even allow alternative APIs to be built that take ABI type definitions to sign directly.

### Outstanding conversations

- EIP-712 original standard includes this comment:
   > Note that the `v` parameter includes the chain id as specified in [EIP-155](https://eips.ethereum.org/EIPS/eip-155).
    - I did not find evidence that this is adopted. Instead, there is strong guidance to use `chainId` in the `EIP712Domain` structure, where it can be verified on-chain using the current `block.chainId`.
    - As a result, the `ethsigner.SignTypedDataV4` code does **not** call `sig.UpdateEIP155(chainID)`

### CODE REVIEW

**Comment by peterbroadhurst** [comment]
@awrichar - looks like there's a DCO issue on the initial commit. Will work with you to get that resolved.

File: pkg/rpcbackend/backend.go [hunk 1]
```diff
@@ -199,8 +199,16 @@ func (rc *RPCClient) SyncRequest(ctx context.Context, rpcReq *RPCRequest) (rpcRe
 	}
 	// JSON/RPC allows errors to be returned with a 200 status code, as well as other status codes
 	if res.IsError() || rpcRes.Error != nil && rpcRes.Error.Code != 0 {
-		log.L(ctx).Errorf("RPC[%s] <-- [%d]: %s", rpcTraceID, res.StatusCode(), rpcRes.Message())
-		err := fmt.Errorf(rpcRes.Message())
+		rpcMsg := rpcRes.Message()
```
Comment (peterbroadhurst) [inline pkg/rpcbackend/backend.go:202]: This was required due to `resty` update

**Comment by awrichar** [comment]
FYI: I went back to add signoff on my commit, then replayed all of @peterbroadhurst 's commits on top. Should be unchanged in terms of the actual code changes.

**Review by nguyer — APPROVED** [review]
Just a quick question about error codes with a possible duplicate, but happy to merge as is

**Pull request merged**